### PR TITLE
TRPL: attributes & conditional compilation

### PR DESCRIPTION
**Pull request by steveklabnik**
r? @alexcrichton

### CODE REVIEW

File: src/doc/trpl/conditional-compilation.md [hunk 1]
```diff
@@ -1,3 +1,66 @@
...
+# fn bar() {}
+
+#[cfg(not(foo))]
+# fn not_foo() {}
+```
+
+These can nest arbitrarily:
+
+```rust
+#[cfg(any(not(unix), all(macos, target_arch = "powerpc")))]
```
Comment (alexcrichton) [inline src/doc/trpl/conditional-compilation.md]: `macos` here should be `target_os = "macos"`

File: src/doc/trpl/conditional-compilation.md [hunk 2]
```diff
@@ -1,3 +1,66 @@
...
+```
+
+When you do this, Cargo passes along a flag to `rustc`:
+
+```
+--cfg feature="${feature_name}"
+```
+
+The sum of these `cfg` flags will determine which ones get activated, and
+therefore, which code gets compiled.
```
Comment (alexcrichton) [inline src/doc/trpl/conditional-compilation.md]: Could you show an example of `#[cfg(feature = "foo")]`?

**Comment by alexcrichton** [comment]
Could this also make a mention of the `cfg!` syntax extension? I would personally recommend using `cfg!` whenever possible over `#[cfg]` as the code will be guaranteed to at least compile on all platforms.

File: src/doc/trpl/conditional-compilation.md (hunk 1 above)
Comment (steveklabnik) [inline src/doc/trpl/conditional-compilation.md]: :+1:

File: src/doc/trpl/conditional-compilation.md (hunk 2 above)
Comment (steveklabnik) [inline src/doc/trpl/conditional-compilation.md]: added sone

**Comment by steveklabnik** [comment]
@alexcrichton i have not actually used `cfg!`. Do we have any examples lying around?

**Comment by alexcrichton** [comment]
There's some example usage in [backtrace.rs](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/backtrace.rs), specifically [here](https://github.com/rust-lang/rust/blob/2baf3482537f5a245a9c17ca730398f1a8b001d7/src/libstd/sys/unix/backtrace.rs#L193) and [here](https://github.com/rust-lang/rust/blob/2baf3482537f5a245a9c17ca730398f1a8b001d7/src/libstd/sys/unix/backtrace.rs#L363-L366)

**Comment by steveklabnik** [comment]
Example added. neat!

**Comment by Ms2ger** [comment]
lgtm

File: src/doc/trpl/conditional-compilation.md [hunk 3]
```diff
@@ -1,3 +1,93 @@
...
+
+```
+--cfg feature="${feature_name}"
+```
+
+The sum of these `cfg` flags will determine which ones get activated, and
+therefore, which code gets compiled. Let’s take this code:
+
+```rust
+#[cfg(feature = foo)]
```
Comment (alexcrichton) [inline src/doc/trpl/conditional-compilation.md]: I think this'll need to be `feature = "foo"` (with the quotes)

**Comment by alexcrichton** [comment]
r=me with a minor nit, thanks!

**Pull request merged**